File: spaced_rep_nov_2022/containerWithMostWater.js
// Two pointer
// Time: O(n), Space: O(1)
 var maxArea = function(height) {
    let maxContainer = -Infinity, s = 0, e = height.length-1

    while (s < e) {
        let area = Math.min(height[s], height[e]) * (e - s)
        maxContainer = Math.max(maxContainer, area)
        height[s] <= height[e] ? s++ : e--;
    }
    return maxContainer;
};
console.log(maxArea([1,8,6,2,5,4,8,3,7])) // = 49

File: swift/containerWithMostWater.swift
class Solution {
    func maxArea(_ height: [Int]) -> Int {
        var area: Int = 0
        var lo: Int = height[0], hi = height[height.count - 1]

        while lo < hi {
            var h = min(height[lo], height[hi]) 
            var w: Int = hi - lo
            area = max(area, h * w)
            lo <= hi ? (lo += 1) : (hi -= 1)
        }
        return area
    }
}